pkg/uploader/provider/restic.go

```diff
@@ -13,22 +13,280 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	log             logrus.FieldLogger
+	bsl             *velerov1api.BackupStorageLocation
+}
+
 func NewResticUploaderProvider(
 	repoIdentifier string,
 	bsl *velerov1api.BackupStorageLocation,
 	credGetter *credentials.CredentialGetter,
 	repoKeySelector *v1.SecretKeySelector,
 	log logrus.FieldLogger,
 ) (Provider, error) {
-	return nil, nil //TODO
+	provider := resticProvider{
+		repoIdentifier: repoIdentifier,
+		log:            log,
+		bsl:            bsl,
+	}
+
+	var err error
+	provider.credentialsFile, err = credGetter.FromFile.Path(repoKeySelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating temp restic credentials file")
+	}
+
+	// if there's a caCert on the ObjectStorage, write it to disk so that it can be passed to restic
+	if bsl.Spec.ObjectStorage != nil && bsl.Spec.ObjectStorage.CACert != nil {
+		provider.caCertFile, err = restic.TempCACertFile(bsl.Spec.ObjectStorage.CACert, bsl.Name, filesystem.NewFileSystem())
+		if err != nil {
+			return nil, errors.Wrap(err, "error create temp cert file")
+		}
+	}
+
+	provider.repoEnv, err = restic.CmdEnv(bsl, credGetter.FromFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "error generating repository cmnd env")
+	}
+
+	return &provider, nil
+}
+
+// Not implement yet
+func (rp *resticProvider) Cancel() {
+}
+
+func (rp *resticProvider) Close(ctx context.Context) error {
+	if err := os.Remove(rp.credentialsFile); err != nil {
+		return errors.Wrapf(err, "failed to remove file %s", rp.credentialsFile)
+	}
+	if err := os.Remove(rp.caCertFile); err != nil {
+		return errors.Wrapf(err, "failed to remove file %s", rp.caCertFile)
+	}
+	return nil
+}
+
+// RunBackup runs a `backup` command and watches the output to provide
+// progress updates to the caller.
+func (rp *resticProvider) RunBackup(
+	ctx context.Context,
+	path string,
+	tags map[string]string,
+	parentSnapshot string,
+	updater uploader.ProgressUpdater) (string, error) {
+	if updater == nil {
+		return "", errors.New("Need to initial backup progress updater first")
+	}
+
+	log := rp.log.WithFields(logrus.Fields{
+		"path":           path,
+		"parentSnapshot": parentSnapshot,
+	})
+
+	// buffers for copying command stdout/err output into
+	stdoutBuf := new(bytes.Buffer)
+	stderrBuf := new(bytes.Buffer)
+
+	// create a channel to signal when to end the goroutine scanning for progress
+	// updates
+	quit := make(chan struct{})
+
+	rp.backupTags = tags
+	backupCmd := ResticBackupFunc(rp.repoIdentifier, rp.credentialsFile, path, tags)
+	backupCmd.Env = rp.repoEnv
+	backupCmd.CACertFile = rp.caCertFile
+
+	if parentSnapshot != "" {
+		backupCmd.ExtraFlags = append(backupCmd.ExtraFlags, fmt.Sprintf("--parent=%s", parentSnapshot))
+	}
+
+	// #4820: restrieve insecureSkipTLSVerify from BSL configuration for
+	// AWS plugin. If nothing is return, that means insecureSkipTLSVerify
+	// is not enable for Restic command.
+	skipTLSRet := restic.GetInsecureSkipTLSVerifyFromBSL(rp.bsl, rp.log)
```

Comment:
Looks like this call can also be moved to `NewResticUploaderProvider`, so that there is no code redundancy in `RunBackup` and `RunRestore`

Reply:
I've changed